db/src/types.rs

```diff
@@ -0,0 +1,251 @@
+pub type Entid = i64;
+
+/// The attribute of each Mentat assertion has an :db/valueType constraining the value to a
+/// particular domain.  Mentat recognizes the following :db/valueType values.
+#[derive(Clone,Debug,Eq,Hash,Ord,PartialOrd,PartialEq)]
+pub enum ValueType {
+    Ref,
+    Boolean,
+    Instant,
+    Long,
+    Double,
+    String,
+    UUID,
+    URI,
+    Keyword,
+}
+
+impl ValueType {
+    pub fn value_type_tag(&self) -> i32 {
+        match *self {
+            ValueType::Ref => 0,
+            ValueType::Boolean => 1,
+            ValueType::Instant => 4,
+            ValueType::Long => 5, // SQLite distinguishes integral from decimal types, allowing long and double to share a tag.
+            ValueType::Double => 5, // SQLite distinguishes integral from decimal types, allowing long and double to share a tag.
+            ValueType::String => 10,
+            ValueType::UUID => 11,
+            ValueType::URI => 12,
+            ValueType::Keyword => 13,
+        }
+    }
+
+    pub fn from_value_type_tag(value_type_tag: &i32) -> Option<ValueType> {
+        match *value_type_tag {
+            0 => Some(ValueType::Ref),
+            1 => Some(ValueType::Boolean),
+            4 => Some(ValueType::Instant),
+            5 => Some(ValueType::Long), // SQLite distinguishes integral from decimal types, allowing long and double to share a tag.
+            // 5 => Some(ValueType::Double),
+            10 => Some(ValueType::String),
+            11 => Some(ValueType::UUID),
+            12 => Some(ValueType::URI),
+            13 => Some(ValueType::Keyword),
+            _ => None
+        }
+    }
+}
+
+/// Represents one partition of the entid space.
+#[derive(Clone,Debug,Eq,Hash,Ord,PartialOrd,PartialEq)]
+pub struct Partition {
+    /// The first entid in the partition.
+    pub start: i64,
+    /// The next entid to be allocated in the partition.
+    pub index: i64,
+}
+
+impl Partition {
+    pub fn new(start: i64, next: i64) -> Partition {
+        assert!(start <= next, "A partition represents a monotonic increasing sequence of entids.");
+        Partition { start: start, index: next }
+    }
+}
+
+/// Map partition names to `Partition` instances.
+pub type PartitionMap = BTreeMap<String, Partition>;
+
+/// A Mentat schema attribute has a value type and several other flags determining how assertions
+/// with the attribute are interpreted.
+///
+/// TODO: consider packing this into a bitfield or similar.
+#[derive(Clone,Debug,Eq,Hash,Ord,PartialOrd,PartialEq)]
+pub struct Attribute {
+    /// The associated value type, i.e., `:db/valueType`?
+    pub value_type: ValueType,
+    /// `true` if this attribute is multi-valued, i.e., it is `:db/cardinality
+    /// :db.cardinality/many`.  `false` if this attribute is single-valued (the default), i.e., it
+    /// is `:db/cardinality :db.cardinality/one`.
+    pub multival: bool,
+    /// `true` if this attribute is unique-value, i.e., it is `:db/unique :db.unique/value`.
```

Comment:
There needs to be a lazy approach here, since one can specify `:db/fulltext true` _before_ `:db/valueType :db.type/string`.  I can think of lots of ways of doing this, but few that are better than manually doing `check_rep` when we create the `Schema` from the `SchemaMap`.  One possibility would be to encapsulate all the changes to the `Attribute` into a builder pattern, sort the builders before application, and then do incremental changes.  It's a lot of work for a little pay off.

For this ticket, I'll type-check on `Schema` construction, and if we need to do more later we can do it.